components/post_editor.tsx

```diff
@@ -0,0 +1,418 @@
+
+    focus() {
+        (this.refs.editor as HTMLDivElement).focus();
+        if (this.selection) {
+            this.setSelectionInEditor();
+        }
+    }
+
+    blur() {
+        (this.refs.editor as HTMLDivElement).blur();
+    }
+
+    componentDidMount() {
+        this.updateEditorHtml(this.props.value || this.props.defaultValue || '');
+    }
+
+    handleInput = () => {
+        this.ensureCorrectNode();
+        this.updateEditorHtmlOnInput();
+        this.currentSelectionInvalid = true;
+        this.setState({showingPlaceholder: ((this.refs.editor as Element).textContent || '') === ''});
+        if (this.props.onInput) {
+            const inputEvent = {
+                target: this,
+            } as any;
+
+            this.props.onInput(inputEvent);
+        }
+    }
+
+    handleKeyDown = (e: React.KeyboardEvent) => {
+        if (Utils.isKeyPressed(e, Constants.KeyCodes.BACKSPACE) ||
+            Utils.isKeyPressed(e, Constants.KeyCodes.LEFT) ||
+            Utils.isKeyPressed(e, Constants.KeyCodes.RIGHT)) {
+            this.currentSelectionInvalid = true;
+        }
+
+        if (this.props.onKeyDown) {
+            this.props.onKeyDown(e);
+        }
+    }
+
+    handlePaste = (e: React.ClipboardEvent<HTMLDivElement>) => {
+        if (!e.nativeEvent.clipboardData) {
+            return;
+        }
+
+        const pastedText = e.nativeEvent.clipboardData.getData('text');
+        const pasteStart = this.selectionStart;
+        const pasteEnd = this.selectionEnd;
+
+        this.value = `${this.value.slice(0, pasteStart)}${pastedText}${this.value.slice(pasteEnd)}`;
+        this.selectionEnd = pasteStart + pastedText.length;
+
+        this.handleInput();
+        e.preventDefault();
+    }
+
+    handleCopy = (e: React.ClipboardEvent<HTMLDivElement>) => {
+        if (!e.nativeEvent.clipboardData) {
+            return;
+        }
+
+        e.nativeEvent.clipboardData.setData('text', this.value);
+        e.preventDefault();
+    }
+
+    render() {
+        const props = {...this.props};
+        const {placeholder} = props;
+
+        Reflect.deleteProperty(props, 'onHeightChange');
+        Reflect.deleteProperty(props, 'providers');
+        Reflect.deleteProperty(props, 'channelId');
+        Reflect.deleteProperty(props, 'emojiMap');
+        Reflect.deleteProperty(props, 'value');
+        Reflect.deleteProperty(props, 'defaultValue');
+        Reflect.deleteProperty(props, 'onInput');
+        Reflect.deleteProperty(props, 'onKeydown');
+        Reflect.deleteProperty(props, 'placeholder');
+
+        const placeholderAriaLabel = placeholder ? placeholder.toLowerCase() : '';
+
+        return (
+            <>
+                {placeholder && this.state.showingPlaceholder && (
+                    <div
+                        {...props}
+                        style={style.placeholder}
+                    >
+                        {placeholder}
+                    </div>)}
+                <div
```

Comment:
I wonder if it wouldn't be easier to have the actual text someplace and generate the nodes from it, so when you do some operation you just have to do it on the text value